### PR TITLE
ATLP-446. Added kerberos in image and validation check for krb user

### PR DESCRIPTION
**Pull request by aklochkova**

### CODE REVIEW

File: src/main/java/com/odysseusinc/arachne/executionengine/service/impl/KerberosServiceImpl.java [hunk 1]
```diff
@@ -37,6 +37,9 @@ public void kinit(DataSourceUnsecuredDTO dataSource, File workDir) throws IOExce
                 CommandBuilder builder = CommandBuilder.newCommand();
                 switch (dataSource.getKrbAuthMethod()) {
                     case PASSWORD:
+                        if (StringUtils.isBlank(dataSource.getKrbUser())) {
```
Comment (wivern) [inline src/main/java/com/odysseusinc/arachne/executionengine/service/impl/KerberosServiceImpl.java]: krbUser is required in both cases, why not to check it once before switch

File: src/main/docker/krb5.conf [hunk 2]
```diff
@@ -0,0 +1,10 @@
+[libdefaults]
+  	  default_realm = ODYSSEUSINC.COM
+
+	[realms]
+	  ODYSSEUSINC.COM = {
+	    admin_server = ovh08.odysseusinc.com
```
Comment (wivern) [inline src/main/docker/krb5.conf]: do we really want to tell everyone on github our server address?

File: src/main/java/com/odysseusinc/arachne/executionengine/service/impl/KerberosServiceImpl.java [hunk 3]
```diff
@@ -53,7 +56,7 @@ public void kinit(DataSourceUnsecuredDTO dataSource, File workDir) throws IOExce
                         try (OutputStream out = new FileOutputStream(keytab.toFile())) {
                             IOUtils.write(dataSource.getKrbKeytab(), out);
                         }
-                        builder.statement("kinit")
+                        builder.statement("C:\\Program Files\\MIT\\Kerberos\\bin\\kinit")
```
Comment (wivern) [inline src/main/java/com/odysseusinc/arachne/executionengine/service/impl/KerberosServiceImpl.java]: path hardcode!!!

Comment (wivern) [reply src/main/java/com/odysseusinc/arachne/executionengine/service/impl/KerberosServiceImpl.java]: Moreover, this should not work inside docker container

File: src/main/java/com/odysseusinc/arachne/executionengine/service/impl/KerberosServiceImpl.java [hunk 4]
```diff
@@ -84,15 +107,50 @@ public void kinit(DataSourceUnsecuredDTO dataSource, File workDir) throws IOExce
...
+        parameters.put("kdcServer", dataSource.getKrbFQDN());
+        context.setVariables(parameters);
+        String htmlString = templateEngine.process(templateName, context);
+
+        Source source = new Source(htmlString);
+        String textString = source.getRenderer().toString();
+
+        File config = new File(configPath);
+        long fileLength = config.length();
+        try (RandomAccessFile raf = new RandomAccessFile(config, "rw"); FileChannel channel = raf.getChannel()) {
```
Comment (wivern) [inline src/main/java/com/odysseusinc/arachne/executionengine/service/impl/KerberosServiceImpl.java]: Potentially [realms] might be not the latest section of krb5.conf, so simply adding realm to the end of file is dangerous.
Why not to add line just under the [realms] ?
And what about concurrency with RandomAccessFile?

File: src/main/java/com/odysseusinc/arachne/executionengine/service/impl/KerberosServiceImpl.java [hunk 5]
```diff
@@ -24,26 +31,39 @@
 
     private static final Logger log = LoggerFactory.getLogger(KerberosService.class);
     public static final String LOG_FILE = "kinit_out.txt";
+    private Map<String, String> krbEnvProps = new HashMap<>();
```
Comment (pavgra) [inline src/main/java/com/odysseusinc/arachne/executionengine/service/impl/KerberosServiceImpl.java]: How is it going to work in multi-threaded env? At least, it should be ThreadLocal, but I'd prefer functional approach

File: src/main/java/com/odysseusinc/arachne/executionengine/service/impl/KerberosServiceImpl.java [hunk 6]
```diff
@@ -66,33 +86,82 @@ public void kinit(DataSourceUnsecuredDTO dataSource, File workDir) throws IOExce
...
-                        .command(command).start();
-                try {
-                    process.waitFor(timeout, TimeUnit.SECONDS);
-                    if (process.exitValue() != 0) {
-                        log.warn("kinit exit code: {}", process.exitValue());
-                    }
-                    process.destroy();
-                    if (log.isDebugEnabled()) {
-                        klist(workDir);
+                if (environmentMode == RuntimeServiceMode.SINGLE) {
```
Comment (pavgra) [inline src/main/java/com/odysseusinc/arachne/executionengine/service/impl/KerberosServiceImpl.java]: Too big for a single function. Also is not atomic for a single function

File: src/main/java/com/odysseusinc/arachne/executionengine/service/impl/KerberosServiceImpl.java [hunk 7]
```diff
@@ -66,33 +86,82 @@ public void kinit(DataSourceUnsecuredDTO dataSource, File workDir) throws IOExce
...
                     }
-                } catch (InterruptedException e) {
-                    log.error("Failed to obtain kerberos ticket", e);
+                } else {
+
+                    String tempKeytabName = keytab.toString();
+                    String commandStr = StringUtils.join(command, " ");
+                    commandStr = commandStr.replace(tempKeytabName, "krb.keytab");
+                    commandStr = commandStr.replace(kinitPath, "");
+                    krbEnvProps.put("KRB_CONF", "/etc/krb5.conf");
```
Comment (pavgra) [inline src/main/java/com/odysseusinc/arachne/executionengine/service/impl/KerberosServiceImpl.java]: I'd proceed with a new (temp) dedicated file for each chroot instance

File: src/main/java/com/odysseusinc/arachne/executionengine/service/impl/RuntimeServiceImpl.java [hunk 8]
```diff
@@ -172,6 +175,9 @@ public void analyze(AnalysisRequestDTO analysis, File file, ResultCallback resul
                     } finally {
                         if (!isExternalJail()) {
                             FileUtils.deleteQuietly(runFile);
+                            for (Path path : tmpPaths) {
```
Comment (wivern) [inline src/main/java/com/odysseusinc/arachne/executionengine/service/impl/RuntimeServiceImpl.java]: I see the tmpPaths is used only to cleanup temporary files, but why not do that outside of this method?
Moreover, it looks like some kind of anti-patterns.

File: src/main/resources/jail.sh [hunk 9]
```diff
@@ -10,17 +10,24 @@ DIST_ARCHIVE=$3
...
+
 #cd $JAIL
 sudo chroot $JAIL /usr/bin/env -i DBMS_USERNAME=$DBMS_USERNAME \
  DBMS_PASSWORD=$DBMS_PASSWORD DBMS_TYPE=$DBMS_TYPE \
  CONNECTION_STRING=$CONNECTION_STRING DBMS_SCHEMA=$DBMS_SCHEMA \
  TARGET_SCHEMA=$TARGET_SCHEMA RESULT_SCHEMA=$RESULT_SCHEMA \
  COHORT_TARGET_TABLE=$COHORT_TARGET_TABLE PATH=$PATH \
  HOME=$HOME IMPALA_DRIVER_PATH=$IMPALA_DRIVER_PATH \
- Rscript /$ANALYSIS_FILE
+ /etc/krb-with-R.sh "$RUN_KINIT" "$ANALYSIS_FILE"
```
Comment (pavgra) [inline src/main/resources/jail.sh]: Why the `krb-with-R.sh` is used always? What if we run in non-kerberos mode? it is going to fail then

File: src/main/java/com/odysseusinc/arachne/executionengine/service/impl/KerberosServiceImpl.java [hunk 10]
```diff
@@ -23,76 +34,175 @@
...
+        parameters.put("realmName", dataSource.getKrbRealm());
+        parameters.put("adminServer", dataSource.getKrbAdminFQDN());
+        parameters.put("kdcServer", dataSource.getKrbFQDN());
+
+        Template confTemplate = TemplateUtils.loadTemplate("templates/krb5Conf.mustache");
+        String textString = confTemplate.apply(parameters);
+
+        File config = new File(configPath);
+        if (isInIsolatedMode) {
+            tempConfigPath.set(Files.createTempFile("", ".conf"));
```
Comment (pavgra) [inline src/main/java/com/odysseusinc/arachne/executionengine/service/impl/KerberosServiceImpl.java]: Pls, use functional approach

**Pull request merged**